### ISSUE TITLE
Pyright incorrectly propagates overloads in higher-order functions

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#8898]
This was originally discovered in https://github.com/microsoft/pyright/issues/8759#issuecomment-2289168344. Please take another look. In my opinion, there's clearly a bug in pyright.

In the following code, `call` is defined in such a way that any function application `f(v)` can be rewritten as `call(f)(lambda: v)`. So the last line can be converted back to just `print(1)`.

```python
from collections.abc import Callable


def call[A, B](f: Callable[[A], B]) -> Callable[[Callable[[], A]], B]:
    return lambda g: f(g())


call(print)(lambda: 1)
```

The above works as expected in pyright. The issue occurs when passing an overloaded function to `call`:

```python
from typing import overload


@overload
def overloaded(v: int) -> None:
    ...


@overload
def overloaded(v: str) -> None:
    ...


def overloaded(v: int | str) -> None:
    pass


call(overloaded)(lambda: 1)
```

The last line can be simplified to `overloaded(1)` and causes no errors at run time. But pyright emits the following error:

```
Argument of type "() -> Literal[1]" cannot be assigned to parameter of type "Overload[() -> int, () -> str]"
  One or more overloads of "" is not assignable
    Type "() -> Literal[1]" is incompatible with type "() -> str"
      Function return type "Literal[1]" is incompatible with type "str"
        "Literal[1]" is incompatible with "str"
```

This happens due to incorrect propagation of overloads when calling higher-order functions.

Revealing the type of `call(overloaded)` gives `(Overload[() -> int, () -> str]) -> None`. Not only the type is inferred incorrectly, but such a definition of an overloaded function itself is illegal since the two cases overlap with incompatible return types. The correct type of `call(overloaded)` should be `Overload[(() -> int) -> None, (() -> str) -> None]`.

**Comment by erictraut** [comment]
Yes, I agree with your analysis. This is indeed a bug.

This will be addressed in the next release.

Interestingly, mypy appears to have a similar bug. It evaluates the type of `call(overloaded)` the same as pyright did prior to the bug fix.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.380.
